### PR TITLE
Tainted bytes

### PR DESCRIPTION
**Pull request by gotcha**
added `TaintedBytes` besides `TaintedString` in `tainted.py`

This is needed to support https://github.com/zopefoundation/Zope/pull/206

### CODE REVIEW

File: src/AccessControl/tainted.py [hunk 1]
```diff
@@ -28,6 +28,24 @@
     from cgi import escape
 
 
+def should_be_tainted(value):
+    if isinstance(value, int):
+        return 60 == value
```
Comment (mauritsvanrees) [inline src/AccessControl/tainted.py]: The 60 comes out of nowhere for me. Could you add a small comment here? Just something like "ord('<') is 60".

File: src/AccessControl/tainted.py [hunk 2]
```diff
@@ -169,3 +187,18 @@ def createOneOptArgWrapper(func):
 
 for f in oneOptArgWrappedMethods:
     setattr(TaintedString, f, createOneOptArgWrapper(f))
+
+
+class TaintedBytes(TaintedString):
+    
+    def __init__(self, value):
+        if isinstance(value, int):
+            value = bytes([value])
```
Comment (mauritsvanrees) [inline src/AccessControl/tainted.py]: Are you sure this works? It gives wildly different results in Python 2 and 3.

Python 3 looks okay:
```
>>> bytes([60, 61, 62])
b'<=>'
```
Python 2 is probably not what we want:
```
>>> bytes([60, 61, 62])
'[60, 61, 62]'
```

Comment (gotcha) [reply src/AccessControl/tainted.py]: Good point. Passing a int should happen only with Python 3.
I'll add a check.

File: src/AccessControl/tainted.py (hunk 1 above)
Comment (gotcha) [reply src/AccessControl/tainted.py]: Yup

File: src/AccessControl/tainted.py (hunk 2 above)
Comment (mauritsvanrees) [reply src/AccessControl/tainted.py]: Won't this fail with an AssertionError exception? And IIRC assert statements might be removed during Python bytecode compiling or with extra optimization flags.

Comment (gotcha) [reply src/AccessControl/tainted.py]: My feeling is that an assert is enough.

File: buildout.cfg [hunk 3]
```diff
@@ -11,7 +11,6 @@ RestrictedPython =
 
 [interpreter]
 recipe = zc.recipe.egg
-interpreter = python
```
Comment (tseaver) [inline buildout.cfg]: What does this change have to do with the pull request?

File: src/AccessControl/tainted.py [hunk 4]
```diff
@@ -169,3 +188,19 @@ def createOneOptArgWrapper(func):
 
 for f in oneOptArgWrappedMethods:
     setattr(TaintedString, f, createOneOptArgWrapper(f))
+
+
+class TaintedBytes(TaintedString):
+    
+    def __init__(self, value):
+        if isinstance(value, int):
+            assert six.PY3
```
Comment (tseaver) [inline src/AccessControl/tainted.py]: Please no assert statements in non-test code:  they get "compiled away" when run with `-O`.

Instead, raise a `ValueError`.

File: src/AccessControl/tests/test_tainted.py [hunk 5]
```diff
@@ -157,3 +180,37 @@ def testStringMethods(self):
...
+class TestTaintedBytes(TestTaintedString):
+
+    def setUp(self):
+        self.unquoted = b'<test attr="&">'
+        self.quoted = b'&lt;test attr=&quot;&amp;&quot;&gt;'
+        self.tainted = self._getClass()(self.unquoted)
+
+    def _getClass(self):
+        from AccessControl.tainted import TaintedBytes
+        return TaintedBytes
```
Comment (tseaver) [inline src/AccessControl/tests/test_tainted.py]: Please add explicit tests for the constructor.

File: buildout.cfg (hunk 3 above)
Comment (gotcha) [reply buildout.cfg]: This causes problem if the buildout is run inside a virtualenv.

**Comment by gotcha** [comment]
@tseaver I adapted the code based on your review. 

Tests do pass on 2.7, 3.5 and 3.6. 
Tests fail on 3.4 for missing support of % for bytes.
Do we need to keep supporting 3.4 ?

File: src/AccessControl/tests/test_tainted.py (hunk 5 above)
Comment (gotcha) [reply src/AccessControl/tests/test_tainted.py]: Done

**Comment by icemac** [comment]
@gotcha If the % operator on bytes is only needed in the tests you could skip them on Python 3.4.
Although no-one should start a new project using Python 3.4 or port a project from 2.7 to 3.4 it is still supported by the Python core developers. We have no consensus yet that we want to drop it before its sunset (seems to be in March 2019 according to PEP 429).

**Comment by icemac** [comment]
Okay it is not only in the test, so maybe it is the time to drop support for Python 3.4 here and in Zope altogether.

**Comment by icemac** [comment]
@tseaver Are you okay with dropping Python 3.4 support to get this PR merged? (It is needed by zopefoundation/Zope#206 which is needed to port zopefoundation/Products.GenericSetup to Python 3)

File: src/AccessControl/tests/test_tainted.py [hunk 6]
```diff
@@ -195,6 +195,7 @@ def _getClass(self):
     
     def testCmp(self):
         self.assertTrue(self.tainted == self.unquoted)
+        self.assertEqual(self.tainted, self.unquoted)
```
[the remark this answers is not in the document]
Comment (gotcha) [reply src/AccessControl/tests/test_tainted.py]: The output of 198 is better imo.

**Comment by icemac** [comment]
@hannosch What do you think about dropping Python 3.4 support because it doesn't support the `%` operator on bytes which is needed for this feature?

**Comment by icemac** [comment]
The only way to make progress with this PR seems to be dropping Python 3.4 support. This PR is needed for zopefoundation/Zope#206 which is needed for zopefoundation/Products.GenericSetup#51.

If there is no veto until 15th of April I am going to merge this PR and drop Python 3.4 support here and in Zope, too.

**Comment by icemac** [comment]
There was no veto in the timeframe I set, so I am going to merge this PR by now.

**Pull request merged**